woocommerce/google-listings-and-ads
# Adjust the infrastructure to uniform the error messages returned from Google Shopping Content service

### Changes proposed in this Pull Request:

This PR is the infrastructure part of #1725 and relates to #1993.

Before making the error message clearer to users, a lot of fundamental changes might be required. Prior to making more changes, I would like to start solution reviews with this PR to make sure this direction is feasible.

Previously, #1258 has pointed out the way to unify GLA error messages for `ApiException` errors thrown from Google Ads APIs. On the Google Shopping Content side, an exception instance may be `Google\Exception` or `Google\Service\Exception`.

The idea is to add the support of these exceptions to the `ExceptionTrait`, renamed from `ApiExceptionTrait`, and then transform the uniformed error messages to `ExceptionWithResponseData` on the error catcher sides.

To do so, this PR:

- Add `get_exception_errors` to `ApiExceptionTrait` as a more general method for getting errors from an exception.
- Rename `ApiExceptionTrait` to `ExceptionTrait` and change all callers to use the general method for getting errors.
- Add the support of `GoogleServiceException` type to `ExceptionTrait::get_exception_errors`.
- Adjust the methods of `MerchantTrait` to mock the possible exceptions.
- Change the relevant getter and updater of Merchant Center account to throw an ExceptionWithResponseData when failed.
- Remove an unused file *src/Exception/MerchantApiException.php*

### Screenshots:

https://github.com/woocommerce/google-listings-and-ads/assets/17420811/c5c03aff-b272-4f6a-bc1d-5f2bf9a422ff

Some other error messages I tested:

![1](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/a7d37f98-e069-4f2c-a37a-57213d3c77a7)
![2](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/1aee199c-231b-435e-afcc-b44a2bf8fe67)

### Detailed test instructions:

1. Disconnect a connected Google Merchant Center account (via the Connection Test page).
2. Go to step 1 of the onboarding flow and connect to the same Google Merchant Center account with an invalid site URL. This can be achieved by changing this line of code locally
   https://github.com/woocommerce/google-listings-and-ads/blob/d0209ea221ba6b8de741e2273ada9c708f9fad76/src/MerchantCenter/AccountService.php#L462
3. It should show a clearer error message in the notification UI than before. Previously, the message was "Unable to update Merchant Center account."
4. Inspect the response body of the `gla/mc/accounts/switch-url` API via Console tab of DevTool. It should be in a format like:
   ```json
   {
       "message": "Unable to update Merchant Center account: [websiteUrl] URL starts with an invalid scheme",
       "errors": {
           "invalid": "[websiteUrl] URL starts with an invalid scheme"
       }
   }
   ```

### Additional details:

About the data structure and format of the `errors` field in the `ExceptionWithResponseData`, it's obviously not very consistent across the different data sources. Since the frontend side is not yet using it but only the `message` field, I think it would be appropriate to continue its optimization when the failed response body returned by GLA APIs is more consistent. The current tuning won't focus much on this part.

### Changelog entry

> Tweak - Make the error message clearer for errors that occur in getting or updating a Google Merchant Center account.
